File: src/main/java/at/searles/buf/BufferedStream.java
package at.searles.buf;

public interface BufferedStream extends CharStream {
    /**
     * @return A pointer that uniquely identifies the
     * next char that is returned by next(). Most likely
     * the index of the underlying code unit like index
     * of charAt() in a string.
     */
    long ptr();

    /**
     * Sets the pointer position to the argument. The underlying
     * implementation must specify the maximum difference between
     * ptr() and the argument here.
     *
     * @param ptr The pointer is reverted
     *            to the corresponding position.
     */
    void setPtr(long ptr);

    class Impl implements BufferedStream {

        private final CharStream stream;
        private final int[] buffer;

        private long ptr;
        private int offset;

        public Impl(CharStream stream, int bufSize) {
            this.stream = stream;
            this.buffer = new int[bufSize];
            this.ptr = 0;
            this.offset = 0;
        }

        public int bufSize() {
            return buffer.length;
        }

        public int cached(long ptr) {
            checkIfPtrInCache(ptr);
            return buffer[(int) (ptr % buffer.length)];
        }

        @Override
        public int next() {
            if (ptr % buffer.length == offset) {
                // next one is a new one.
                int ch = stream.next();

                if (ch == -1) {
                    return -1;
                }

                buffer[offset] = ch;
                offset = (offset + 1) % buffer.length;
                ptr++;

                return ch;
            }

            return buffer[(int) (ptr++ % buffer.length)];
        }

        @Override
        public long ptr() {
            return ptr;
        }

        private void checkIfPtrInCache(long ptr) {
            if (this.ptr < ptr) {
                throw new IllegalArgumentException("cannot foresee future");
            }

            if (this.ptr - ptr >= buffer.length) {
                throw new IllegalArgumentException("buffer is too small");
            }
        }

        @Override
        public void setPtr(long ptr) {
            checkIfPtrInCache(ptr);
            this.ptr = ptr;
        }
    }
}

File: src/main/java/at/searles/buf/BufferedStream.kt
package at.searles.buf

interface BufferedStream : CharStream {
    /**
     * @return A pointer that uniquely identifies the
     * next char that is returned by next(). Most likely
     * the index of the underlying code unit like index
     * of charAt() in a string.
     */
    var position: Long

    class Impl(private val stream: CharStream, bufSize: Int) : BufferedStream {
        private val buffer: IntArray = IntArray(bufSize)
        private var bufferPtr: Int = 0
        private var ptr: Long = 0

        override var position: Long
            get() = ptr
            set(value) {
                checkIfPtrInCache(value)
                ptr = value
            }


        fun bufSize(): Int {
            return buffer.size
        }

        fun cached(ptr: Long): Int {
            checkIfPtrInCache(ptr)
            return buffer[(ptr % buffer.size).toInt()]
        }

        override fun next(): Int {
            if (ptr % buffer.size != bufferPtr.toLong()) {
                return buffer[(ptr++ % buffer.size).toInt()]
            }

            // next one is a new one.
            val ch = stream.next()

            if (ch == -1) {
                return -1
            }

            buffer[bufferPtr] = ch
            bufferPtr = (bufferPtr + 1) % buffer.size
            ptr++
            return ch
        }

        private fun checkIfPtrInCache(ptr: Long) {
            require(this.position >= ptr) { "cannot foresee future" }
            require(this.position - ptr < buffer.size) { "buffer is too small, Maximum size is ${buffer.size}" }
        }

    }
}